flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1772,6 +1810,21 @@ class ScalarExprLowering {
       }
 
       if (arg.passBy == PassBy::MutableBox) {
+        if (Fortran::evaluate::UnwrapExpr<Fortran::evaluate::NullPointer>(
+                *expr)) {
+          // If expr is NULL(), the mutableBox created must be a deallocated
+          // pointer with the dummy argument characteristics (see table 16.5
+          // in Fortran 2018 standard).
+          auto boxTy = fir::dyn_cast_ptrEleTy(argTy);
+          assert(boxTy && boxTy.isa<fir::BoxType>() &&
+                 "must be a fir.box type");
+          auto boxStorage = builder.createTemporary(loc, boxTy);
+          auto nullBox = Fortran::lower::createUnallocatedBox(
+              builder, loc, boxTy, /*nonDeferredParams=*/{});
```

Comment:
Right, no length parameters are set for the created box because any non deferred type parameters of the dummy will be evaluated on the callee side, and it is illegal to use NULL without a MOLD if any dummy length parameters are assumed. (added verbatim as a comment).

> Side question: can/does semantics check that condition?

I just checked, and no. The front-end could enforce this specific constraint, but in general, I believe it is illegal to pass di-associated pointers with deferred type parameters  that are assumed in the dummy argument type, and there is no way to check that at compile time. So this case still falls under a general user constraint, though in this case, it could be enforced at compile time.